### ISSUE TITLE
AutoRelay/RelayService

### ISSUE THREAD

**Issue by QZAiXH** [hyprspace/hyprspace#78]
I am a newbie and I want to try to modify the code.
![image](https://user-images.githubusercontent.com/23068780/165669050-599dec4b-6ae2-49b6-a6fe-a1c14b2dcac8.png)

I have three nodes, A B C. When A and B are behind NAT and I want to relay via C, I add AutoRelay, RelayService to libp2p.new, but it doesn't work

**Comment by Winterhuman** [comment]
AutoRelay (CircuitV1) is deprecated, use RelayClient (CircuitV2) instead and that *should* work

**Comment by QZAiXH** [comment]
Hi, thank you for your reply. I didn't find the RelayClient option, where is it.
I am using go-libp2p 0.19.0, Isn't AutoRelay the v2 used?

**Comment by QZAiXH** [comment]
Do I need to specifically specify a relay node and will libp2p automatically post the relay address for me?

**Comment by Winterhuman** [comment]
Ah sorry, I got that wrong. AutoRelay *used* to be for CircuitV1, but now all that stuff has been removed (https://github.com/libp2p/go-libp2p/pull/1368/commits/29fd1022a490acc1a155a78170f09729b57c7aa5).

I'm honestly sure where to point you, Libp2p haven't updated the docs or provided any CircuitV2 examples as far as I know, you might want to search for the term "CircuitV2" in GitHub and see if you find any projects that use it.

**Comment by alecbcs** [comment]
@QZAiXH I'll take a look into CircuitV2 this weekend and what would need to be implemented to enable this. So far I've held off on integrating a circuit relay into hyprspace because in version 1 there weren't good authentication controls. Within IPFS this meant that if you enabled the relay on one of your nodes, *anyone on the entire IPFS network* could use your node to relay data which could have quickly ate up many people's bandwidth and thus it was highly discouraged that anyone ran their own circuit relay. Instead trusted relays were provided by the IPFS team.

I'm curious about your use case, Hyprspace should already be able to connect your two peers (A & B) that are both behind NATs without having to go through a third peer. Have you tested that yet and if so are your firewalls blocking anything in specific that we could work on?

**Comment by QZAiXH** [comment]
Yes, I performed the test. I tested between my company computer and my friend's company computer and we could not connect to each other. My company's NAT type is Port Restricted Cone NAT, my friend's is unknown for now

**Comment by QZAiXH** [comment]
Hi @alecbcs, any progress on CircuitV2 please? I have researched the code for ipfs and it just creates CircuitV2, how to limit the users connected is not seen.

**Comment by alecbcs** [comment]
@QZAiXH is there any chance that your NAT blocks access to non-standard ports? (80, 443, etc...) I'd be curious if you all are able to connect if you do something similar to #80.

Also apologies on the bad communication around the delay getting circuit relay into Hyprspace, I started working on the circuit relay integration but that lead rise to a much larger discussion of a rewrite of the way that we handle peers in general/how we should allow people to create private hyperspace clusters with their own bootstraping peers. Anyway it's coming but I don't have a estimated date yet. In the mean time I'm curious if #80 gives us any insight into your situation.

**Comment by QZAiXH** [comment]
My friend's company network is Symmetric NAT, and port 80 443 is already occupied, so I can only use a relay to connect.

**Comment by alecbcs** [comment]
@QZAiXH I want to thank you for your interest in Hyprspace, but if you require a relay to connect between the two nodes why not just use Wireguard to connect the nodes instead? Hyprspace is really meant to connect two nodes directly without passing packets between a publicly available node like as is the case with Wireguard.

**Comment by QZAiXH** [comment]
@alecbcs thanks for your suggestion. However, what I mentioned above are just my individual scenarios. What I want to achieve is that it is best to connect directly, and relay if not. I think Hyprspace can do exactly the same as zerotier and tailscale.
